File: TVTComment/Model/IPC/IPCMessage/CurrentChannelIPCMessage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TVTComment.Model.IPC.IPCMessage
{
    class CurrentChannelIPCMessage : IIPCMessage
    {
        public ChannelInfo Channel;
        public EventInfo Event;

        public string MessageName => "CurrentChannel";

        public void Decode(IEnumerable<string> content)
        {
            string[] contents = content.ToArray();
            if (contents.Length != 16)
                throw new IPCMessageDecodeException("CurrentChannelのcontentの数が16以外です");

            Channel = new ChannelInfo();
            try
            {
                Channel.SpaceIndex = int.Parse(contents[0]);
                Channel.ChannelIndex = int.Parse(contents[1]);
                Channel.RemoteControlKeyId = int.Parse(contents[2]);

                Channel.NetworkId = ushort.Parse(contents[3]);
                Channel.TransportStreamId = ushort.Parse(contents[4]);
                Channel.ServiceId = ushort.Parse(contents[5]);
                
                Channel.NetworkName = contents[7];
                Channel.TransportStreamName = contents[8];
                Channel.ServiceName = contents[9];
                Channel.ChannelName = contents[10];

                Event = new EventInfo(ushort.Parse(contents[6]), contents[11], contents[12], contents[13],
                    DateTime.SpecifyKind(DateTimeOffset.FromUnixTimeSeconds(long.Parse(contents[14])).DateTime, DateTimeKind.Local),
                    new TimeSpan(0, 0, int.Parse(contents[15])));
            }
            catch(FormatException)
            {
                throw new IPCMessageDecodeException("CurrentChannelのフォーマットが不正です");
            }
        }

        public IEnumerable<string> Encode()
        {
            throw new NotImplementedException("CurrentChannelIPCMessage::Encodeは実装されていません");
        }
    }
}

File: Viewer/TVTComment/IPC/IPCMessage/CurrentChannelIPCMessage.cpp
#include "stdafx.h"
#include "CurrentChannelIPCMessage.h"

namespace TVTComment
{
	std::string CurrentChannelIPCMessage::GetMessageName() const
	{
		return u8"CurrentChannel";
	}

	std::vector<std::string> CurrentChannelIPCMessage::Encode() const
	{
		std::vector<std::string> ret;

		ret.push_back(std::to_string(this->SpaceIndex));
		ret.push_back(std::to_string(this->ChannelIndex));
		ret.push_back(std::to_string(this->RemotecontrolkeyId));

		ret.push_back(std::to_string(this->NetworkId));
		ret.push_back(std::to_string(this->TransportstreamId));
		ret.push_back(std::to_string(this->ServiceId));
		ret.push_back(std::to_string(this->EventId));

		ret.push_back(this->NetworkName);
		ret.push_back(this->TransportstreamName);
		ret.push_back(this->ServiceName);
		ret.push_back(this->ChannelName);

		ret.push_back(this->EventName);
		ret.push_back(this->EventText);
		ret.push_back(this->EventExtText);
		ret.push_back(std::to_string(this->StartTime));
		ret.push_back(std::to_string(this->Duration));

		return ret;
	}

	void CurrentChannelIPCMessage::Decode(const std::vector<std::string> &)
	{
		throw std::logic_error("CurrentChannelIPCMessage::Decode is not implemented");
	}
}